FEATURE: Inline editing overhaul

Sebobo opened the pull request:
**resolves:**
Resolves: https://github.com/neos/neos-ui/issues/3972
Resolves: https://github.com/neos/neos-ui/issues/3977
Resolves: https://github.com/neos/neos-ui/issues/3123
Resolves: https://github.com/neos/neos-ui/issues/3727
Resolves: https://github.com/neos/neos-ui/issues/3853

Resolves: https://github.com/neos/neos-ui/issues/3039 (we no longer replicate the ui)
Resolves: https://github.com/neos/neos-ui/issues/3512 (we no longer replicate the ui)

**replaces: (because incorporated into this mega pr)**
https://github.com/neos/neos-ui/pull/4001
https://github.com/neos/neos-ui/pull/3978
https://github.com/neos/neos-ui/pull/3979



**checklist**

- [x] Fully integrate CKeditor block and inline toolbars with editable properties
- [x] Remove React wrapping code from Neos.Ui for all editors (except link editor)
- [x] Cursor position dependent "Add" and "Paste" buttons
- [x] Position aware paste which allows skipping the insertion mode modal
- [x] Custom styles / CSS classes selection via yaml configuration and CK plugin
- [x] Inspector in full screen mode
- [x] New breadcrumb that replaces "Current Selection"
- [x] Polyfill for Firefox and older Safari version to support anchor positioning -> custom fallback implementation, as the polyfill has various issues and drawbacks
- [x] Duplicate node button
- [x] Integrate CKEditor toolbars with RichTextEditor 
- [x] Inline table editor with caption support
- [x] Still support old style editor plugins
- [x] Improved content collection overlay
- [x] Integrate and adjust CKEditor theme
- [x] Style detail adjustments
- [ ] Toolbar configuration via YAML (not urgent, maybe for 9.2)
- [x] New "light" theme for inline toolbars
- [x] Various additional plugins enabled (undo, horizontalLine)
- [x] Update CKEditor from 44 to 47 
  - Improved performance on initial, loading and saving & bugfixes
  - We can now fully expose the api for plugins
  - General performance improvements 15-45% lower editor load times
  - Features: Emoji dropdown, fullscreen editor, custom icons
  - Custom icon support
  - Typing improvements especially for Safari browsers
  - Improved pasting tables from word documents
  - Improved "remove formatting"
  - List marker styling
  - Improved markdown conversion

Sebobo committed:
TASK: Simplify node toolbar CSS with flexbox

This gets rid of the float magic

Sebobo committed:
BUGFIX: style and size for IconButton was shown as required though they are optional

Sebobo committed:
FEATURE: Drag and drop nodes in the guest frame

Sebobo committed:
BUGFIX: Edittoolbar style didn’t fully apply due to typo in classname

Sebobo committed:
TASK: Use flexbox for inline toolbars

Sebobo committed:
FEATURE: Move inline editing toolbar into nodetoolbar

With this change we also a CSS anchor to allow us reponsive behaviour of the toolbar depending on its own and the window size

Sebobo committed:
BUGFIX: Move ckeditor body collection wrapper into guest frame

The inline editor uses various elements for highlights which don’t work in the host frame.
With this patch, we instead attach the collection to the guestframe body.

Sebobo committed:
FEATURE: Show ckeditor drag & drop indicators

With this change, the user can see an indicator where selected elements in the inline editor would show up.
The drag & drop feature worked before, but this makes the feature much more usable and intuitive.
It even works between multiple editor instances.

Resolves: #3977

Sebobo committed:
TASK: Set patch folder for yarn to the folder we actually use

Sebobo committed:
BUGFIX: Let ckeditor ignore neos specific drag events

CKEditor catches all drag events so we cannot handle them ourselves.
With this patch it ignores the ones with a neos type

Sebobo committed:
BUGFIX: Reattach ckeditor bodycollection after guestframe reload

Sebobo committed:
TASK: Improve drop indicator layout

Sebobo committed:
TASK: Add missing methods to NodeTypesRegistry interface

Sebobo committed:
TASK: Cleanup custom bodycollection code

Sebobo committed:
TASK: Move dragndrop functionality into a single guest frame component

Sebobo committed:
FEATURE: Use CK formatting toolbar instead of the custom implementation

With this change we also include the CKEditor theme, which we control via CSS variables.

Sebobo committed:
BUGFIX: Make nodetoolbar zindex work again

Sebobo committed:
FEATURE: Improved empty collection overlay

With this change the empty collection overlay is always visible and
shows a label and an indicator so it’s easier to select and add content

Sebobo committed:
FEATURE: Split node toolbar buttons into a dropdown

Sebobo committed:
TASK: Rearrange node toolbar buttons

Sebobo committed:
FEATURE: Move right handed secondary toolbar actions to primary

Sebobo committed:
FEATURE: Support insert position for paste node

Sebobo committed:
FEATURE: Replace „Selected element“ with breadcrumb in secondary toolbar

Sebobo committed:
FEATURE: Move EditorToolbar to SecondaryToolbar

Sebobo committed:
TASK: Improve button and collection styling

Sebobo committed:
TASK: Refactor inline node toolbars

Sebobo committed:
FEATURE: Add custom style support in inline editor

Sebobo committed:
TASK: Convert node toolbars to functional components

This reduces rerendering to a minimal amount

Sebobo committed:
TASK: Improve breadcrumb styling

Sebobo committed:
TASK: Adjust structural toolbar css classnames

Sebobo committed:
TASK: Small context path cleanup

Sebobo committed:
TASK: Use fixed CKeditor versions

Sebobo committed:
FEATURE: Make CK and other node toolbars usable at the same time

With this change the structural toolbar is anchor to the context toolbar to prevent them from overlapping.
Also by default the context toolbar is in front and only the CK toolbar is in front if it is hovered.

Sebobo committed:
TASK: Move preview and fullscreen buttons back to secondary toolbar

Sebobo committed:
FEATURE: Allow inspector in fullscreen mode

Sebobo committed:
TASK: Adjust ck dependency versions

Sebobo committed:
TASK: Limit breadcrumb to current document

Sebobo committed:
BUGFIX: Show inspector toggle in fullscreen mode

Sebobo committed:
BUGFIX: Resolve some case of outline collisions between editables and the focused node

If the editable is the outer element of a rendered node there was a collision between
the outline of the focusedNode class and the styles that CKEditor sets.

Sebobo committed:
TASK: Remove unused svgToDataUri

This was only used for the table editor, which was replaced by the native CK table editor.

Sebobo committed:
TASK: Adjust drag button to new toolbar layout

Sebobo committed:
TASK: Improve CK toolbar styling

Sebobo committed:
TASK: Use new hook to access redux store

Sebobo committed:
TASK: Remove ck inline editor border

Sebobo committed:
FEATURE: Enable table caption plugin

Sebobo committed:
FEATURE: Make editable area visible without visual jumping due to border

Sebobo committed:
TASK: Allow selectable CK styles in formatting proptypes

Sebobo committed:
TASK: Reduce visual clutter in breadcrumb

Sebobo committed:
TASK: Make empty collection overlay lighter

Sebobo committed:
BUGFIX: Prevent unfocus events during inline editing

Keep the Neos UI and CKEditor from unfocusing our currently edited node
or editable area when we click parts of the toolbar.

Sebobo committed:
BUGFIX: Prevent delayed repositioning of toolbars during inline editing

Before it would take some time during editing when newlines are added until
the anchorposition would be recalculated due to the stored node change.

Sebobo committed:
TASK: Make CK undo plugin optional

Sebobo committed:
BUGFIX: CK table editor styling

Sebobo committed:
BUGFIX: Show new ckeditor toolbar in RichTextEditor

Sebobo committed:
TASK: Adjust imports to code changes in parent branch

Sebobo committed:
FEATURE: Duplicate node via node context menu

Sebobo committed:
TASK: Resolve linting issues

Sebobo committed:
BUGFIX: Show unselected grid elements in table creation popup

Sebobo committed:
TASK: Use translation singleton in duplicate node button

Sebobo committed:
TASK: Switch context toolbar to dark theme

Sebobo committed:
TASK: Improve CKEditor style selection styling

Sebobo committed:
TASK: Add border to node context menu dropdown

Sebobo committed:
BUGFIX: Use correct font in node label of context toolbar

Sebobo committed:
TASK: Remove deprecated plugins

Sebobo committed:
FEATURE: Allow horizontal line in ckeditor

Sebobo committed:
FEATURE: Update ckeditor from v44 to v47

With this change we are now able to also fully expose the core parts of
the ckeditor api instead of specific classes. Additional modules
can be exposed in further releases.

Sebobo committed:
BUGFIX: Skip insertmode selection on paste in guest frame

Sebobo committed:
TASK: Add shims for ck api exposure

Sebobo committed:
TASK: Update ckeditor api exposure comment

Sebobo committed:
TASK: Adjust manual ckeditor test to editor changes

Sebobo committed:
Merge branch '9.1' into feature/3972-node-toolbars

Sebobo committed:
Merge branch 'feature/3972-node-toolbars' into feature/3972-ckeditor-update

Sebobo committed:
FEATURE: Light ckeditor toolbar theme and improved layout behaviour

mhsdesign committed:
TASK: Add e2e test for legacy `ckeditor5-exports` api

Sebobo committed:
TASK: Cleanup ckeditor creation

Sebobo committed:
Merge pull request #4028 from neos/feature/3972-ckeditor-update

FEATURE: Update ckeditor from v44 to v47

Sebobo committed:
BUGFIX: Toolbar positioning fallback for Firefox and older Safari versions

Sebobo committed:
Merge branch '9.1' into feature/3972-node-toolbars

Sebobo committed:
TASK: Solve linting issues

Sebobo committed:
Merge branch '9.1' into feature/3972-node-toolbars

mhsdesign commented:
bugs i noticed

| Description | Browsers | Fixed | Screenshot |
|--------|--------|--------|--------|
| Inline toolbar not accessible when create after visible | Saf18.5, Edg142.0 |  | <img width="173" height="134" alt="image" src="https://github.com/user-attachments/assets/b976efe3-e100-4f3b-9d1b-91ce22356826" /> |
| Drag indicator is shown for document node (when document node is wrapped? Found in the Neos Ui testing distribution) | Edg142.0 |  | <img width="104" height="126" alt="image" src="https://github.com/user-attachments/assets/87aab001-243b-40cf-8a0d-04d6d3263558" /> |
| Full width content elements cause horizontal scrollbar due to additional pixel when selected | Edg142.0 |  |  <img width="76" height="131" alt="image" src="https://github.com/user-attachments/assets/33e7fe34-ef3f-4da1-8df0-fdc44b0d2355" /> |
|  `currentEditor.execute()` throws nullpointer, legacy toolbar buttons only work after using new toolbar at least once  | Edg142.0 | | <img width="86" height="49" alt="image" src="https://github.com/user-attachments/assets/d636ef47-c804-432c-bcbd-0485633b4207" /> |
| Animated hover insert arrow in columns behaves odd as transform is reset when clicking button, remove animation instead | Edg142.0 | | <img width="72" height="45" alt="image" src="https://github.com/user-attachments/assets/3ba3c605-cbf2-4ff4-bf21-9218a049e952" /> | 
| For text contents from demo import the ckeditor border is smaller than the text (caused by a "wrong"? `<h4>` tag)  | **only** Saf18.5 | | <img width="214" height="94" alt="image" src="https://github.com/user-attachments/assets/e5f8a220-89f5-4e27-b6a0-fef219a73846" /> | 
| Edited text in focus shows new tight grey border - either keep the blue offset one or remove in focus state? | Edg142.0 | | <img width="147" height="61" alt="image" src="https://github.com/user-attachments/assets/4322dd06-225e-4c22-8608-65378aff8054" /> | 
| Clicking on the current entry in the breadcrumb shows mouse pointer and hover but no feedback as this is a noop this makes users cry | | | <img width="205" height="53" alt="image" src="https://github.com/user-attachments/assets/7a6da779-609f-4481-ab8d-0589e57de504" /> | 
| Description | Browsers | | Screenshot |

Sebobo committed:
TASK: Align ckeditor toolbar in secondary editor below editor toolbar with full width

Sebobo committed:
BUGFIX: Store last active editor for editor toolbar buttons

Sebobo committed:
TASK: Group ckeditor format options

Sebobo committed:
TASK: Move contextual node actions into one line again

mhsdesign committed:
PATCH: Use `LinkEditing` plugin instead of ck `Link` plugin which brings an inline Ui

these things are now also grouped in a dedicated Plugin with `requires()` as they are to be used as single unit.

Sebobo committed:
BUGFIX: Reused domnode warning by ckeditor

Sebobo committed:
BUGFIX: Sort structuraltoolbar when below behind ckeditor balloon toolbar

Sebobo committed:
TASK: Resolve linting issues

Sebobo committed:
TASK: Resolve layout issues with context menu dropdown

mhsdesign committed:
BUGFIX: CKEditor focus tracker must not use host frame `document`

We cannot access "document" as this will always be the host frame.
If ckeditor is used in the guest iframe we need to use their document.

The major problem here is the `_blur` method.
Technically relying on the first "ownerDocument" of the possibly multiple elements is not correct as they _could_ belong to different documents but in practice they dont and should not.

Without that patch in the `_blur` method the buttons in the toolbar of the ckeditor could not be toggled correctly. Instead of getting focus on the button on pressing the button and closing the dropdown on release, the focus already lead to closing the dropdown and the mouse release - if on that button - would reopen it again leading to a buggy behaviour. Any dropdown like the paragraph dropdown or to show more options was affected.

Sebobo committed:
TASK: Simplify ck toolbar injection

mhsdesign committed:
BUGFIX: CKEditor tooltip manager must not track mouse in host frame `document`

We need to track the mouse position IN the iframe where the CKEditor is mounted as the document itself is the host frame.
This patch fully opts in to only support the iframe. CKEditor mounted in the host frame - for example the RTE Editor will not show tooltips.
That is the bargain. To fix both cases we need to actually boot and load CKEditor in the iframe by using code sharing.

We cannot support both cases via a simple patch as TooltipManager uses as singleton pattern. We would need to change multiple parts and logic.

mhsdesign committed:
TASK: Adjust manual ckeditor test to use iframe as well

this setup shows more accurately what problems neos has.

Sebobo committed:
TASK: Crop nodelabel via css

Sebobo committed:
TASK: Reenable dropdown for ckeditor toolbar

Sebobo committed:
TASK: Scroll to selected element when clicking in breadcrumb

Sebobo committed:
TASK: Remove empty collection overlay animation

Sebobo commented:
@mhsdesign 

- [x] Inline toolbar not accessible when create after visible
- [x] Drag indicator is shown for document node (when document node is wrapped? Found in the Neos Ui testing distribution) -> weird as I check the role of the node when verifying if it can be dragged
- [ ] Full width content elements cause horizontal scrollbar due to additional pixel when selected -> need to check
- [x] currentEditor.execute() throws nullpointer, legacy toolbar buttons only work after using new toolbar at least once
- [x] Animated hover insert arrow in columns behaves odd as transform is reset when clicking button, remove animation instead -> Removed
- [x] For text contents from demo import the ckeditor border is smaller than the text (caused by a "wrong"? `<h4>` tag) -> editable with `block = false` are smaller as they are just a span inside the h4 f.e.
- [ ] Edited text in focus shows new tight grey border - use outline for hover and active state instead of border for the later
- [x] Clicking on the current entry in the breadcrumb shows mouse pointer and hover but no feedback as this is a noop this makes users cry -> Requires a new label component that behaves like the IconButton, so it looks good in the breadcrumb, but I now enabled the scrollIntoView for each breadcrumb item, this way it actually makes sense to have it as a button.
- [x] --collection-label: "Add content"; needs to be translated via data attribute or similar
- [x] Make the neos variables available with a neos namespace in the guest frame so we can use them here eigentlich -> let's introduce css variable usage with 9.2 (todo: Link issue)
- [x] Check whether we still need the "Extensible style selector" -> now deprecated and we have a replacement with the native ckeditor style selector
- [x] Check if `useState<DecoupledEditor>()` is an issue and can be solved differently
- [x] Replace `formatNodeLabel` with a tested helper function
- [x] Check remaining todo's and fixme's

mhsdesign committed:
FEATURE: Implement ballon edit link, and unlink buttons like in `LinkUI` from `@ckeditor/ckeditor5-link`

mhsdesign committed:
FEATURE: Open new link editor dialog from LinkUi CKEditor plugin and apply change - reintroduce CTRL+K

mhsdesign committed:
TASK: Move inline link handling to `neos-ui-link-editor-inline` and integrate cleanly

we no longer rely on the global neos blob state in `this.editor.neos` and also no longer need to expose the neos link editor instance globally but pass it as factory argument

mhsdesign committed:
PATCH: Apply neos code conventions to LinkUiPlugin

mhsdesign committed:
PATCH: Hide new link editor ballon toolbar when clicking into other editor, inside of iframe or into Neos host

To hide the toolbar when clicking outside the editor we use the focusTrackerInspired from the Balloontoolbar

https://github.com/ckeditor/ckeditor5/blob/v47.2.0/packages/ckeditor5-ui/src/toolbar/balloon/balloontoolbar.ts

This is new as the actual link editor UI does not use the focus tracker this way - and actually HAS problems when being used in the Neos Ui

mhsdesign committed:
PATCH: Cleanup too much whitespace

mhsdesign committed:
PATCH: Move `@ckeditor/ckeditor5-highlight` dependency declaration to correct package

mhsdesign committed:
PATCH: Move remove obsolete `store` dependency in `createEditor()`

mhsdesign committed:
PATCH: Simplify `addPlugin()` function to remove "LEGACY" comment

the single special case is now inlined ;)

mhsdesign committed:
TASK: Migrate CKEditor configuration / integration to typescript as CKEditor supports typescript now ;)

Sebobo committed:
TASK: Remove irrelevant todos and fixmes

Sebobo committed:
TASK: Use ref instead of state for editor in ckwrapper

Sebobo committed:
TASK: Simplify node label generation in toolbar

Sebobo committed:
TASK: Update ckeditor clipboard patch

Sebobo committed:
TASK: Simplify code in node toolbar

Sebobo committed:
FEATURE: Localise empty collection label

Sebobo committed:
TASK: Deprecate generic style selector

This was an undocumented feature and instead one can now use ckeditor native style dropdown which is configurable via yaml.

Sebobo committed:
BUGFIX: Add fallback position for structural toolbar when displayed at the bottom of the viewport

mhsdesign committed:
PATCH: Pacify linter

mhsdesign committed:
PATCH: Improve hide new link editor ballon toolbar when clicking into other editor, inside of iframe or into Neos host

This updates commit 4aacdf4cd9051fbda2ad1ad4939a15d932525fa2.

Previously clicking into another editor did only work on safari and due to retacking the focus `updateFocus=true` it was bugged

mhsdesign committed:
PATCH: Dont focus inline UI when creating new link via toolbar

otherwise in the background of the modal one would see the inline link ui toolbar which is unreachable and it buggs a little the focus handling

mhsdesign committed:
PATCH: Simplify Link Ui `_showUi` and dont open modal on double click

double click is sometimes confusing as its also interpreted as such when selecting one link and then clicking on another link = opens dialog

the shortcut handling was also simplified as we now always go directly to the modal `_handleLinkEditing`

mhsdesign committed:
PATCH: Remove link fake visual selection which is obsolete as we show a big modal and nothing is to be seen

Ckeditor needs this to click the link plugin on text that will be a link so the users knows what text was selected before.

The `LinkEditing` also adds highlighting via `inlineHighlight(editor, 'linkHref', 'a', HIGHLIGHT_CLASS);` to all actual links.

mhsdesign committed:
BUGFIX: Prevent content node to lose focused blue outline when being a ckeditor mount point

Clicking into the host frame or opening the link editor modal (and closing) leaves the element still focused to Neos but its no longer visually indicated.

This is because CKEditor seems to reset the dom element to its previously known state when defocusing during adding `.ck-blurred` to the element in the `EditableUIView`
To prevent a hickup here we hack our ways around and get ahold the ckeditor instance - if any and then use the same way the `EditableUIView` sets  `.ck-blurred` and  `.ck-focused` onto the element by using the view writer

Its a little bit hacky to use the `ckeditorInstance` here as this couples the neos-ui-guest-frame to the ckeditor package ... maybe well find another ways some time
-> An alternative would be to patch out in `EditableUIView` to set the `.ck-blurred` class this will keep the element alive

Sebobo committed:
TASK: Use single quotes in js via editorconfig

Sebobo committed:
TASK: Import svg icon from font awesome

Sebobo committed:
BUGFIX: Update hover state for parent nodes

Resolves: #3123

Sebobo committed:
TASK: Highlight nodelabel in context toolbar on hover

Sebobo committed:
BUGFIX: Adjust placeholder override for latest ckeditor styles

Sebobo committed:
BUGFIX: Ignore focus tracking in ckeditor when dragging nodes

Sebobo committed:
BUGFIX: Hide dropindicator initially

Sebobo committed:
BUGFIX: Structural toolbar fallback for bottom position

Sebobo committed:
BUGFIX: Observe guest frame window resizing to position node toolbar

Sebobo committed:
TASK: Resolve context menu issues with firefox

Sebobo committed:
TASK: Toggle hide/show button caption

Sebobo committed:
TASK: Merge contextmenu labe and button to improve behaviour

Sebobo committed:
TASK: Resolve linting issue

Sebobo committed:
TASK: Update translations

Sebobo committed:
TASK: Solve linting error

Sebobo committed:
Merge branch '9.1' into feature/3972-node-toolbars

mhsdesign committed:
FEATURE: Show outline while editing text

CKEditor natively adds the `.ck-focused` class to active editables.
It also styles that by using a greyish css border.
With the new inline overhaul we load this style but it does not fit with the known dotted blue outline which is visible beforehand on hovering. The ckeditor border is positioned tighter and also due to the difference in color the behaviour when clicking into editables feels wrong.

Now instead of fully disabling the border we want to introduce one ourselves as its a missing feature

see https://github.com/neos/neos-ui/issues/3853

The dottet border known during hover will now be kept during editing a node. There is one exception: When the node only has one html element and that is the ckeditor editable we cannot show two outlines as there is just one element. For these simple editables we omit the dotted outline and instead only show the solid blue focused node outline.

mhsdesign committed:
PATCH: Fix `cut` icon in inline toolbar

mhsdesign committed:
Merge remote-tracking branch 'origin/9.1' into feature/3972-node-toolbars

Sebobo committed:
TASK: Make toolbar anchor position compatible with Firefox 147

Sebobo committed:
TASK: Adjust toolbar ids

Sebobo committed:
Merge branch '9.1' into feature/3972-node-toolbars

mhsdesign committed:
PATCH: Fix duplication inline ui test by switching to iframe when necessary

mhsdesign reviewed (approved):
All todos are fixed what is left are optional things which are extracted to #4046 

e2e ci is red because the tests are run against 9.0 for some reason (pull_request_target is bugged) now i ran them locally where they are fine.

mhsdesign committed:
PATCH: Små cosmetic polish before merge
